go/vt/srvtopo/keyspace_filtering_server_test.go

```diff
@@ -0,0 +1,218 @@
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package srvtopo
+
+import (
+	"fmt"
+	"reflect"
+	"sync"
+	"testing"
+
+	"golang.org/x/net/context"
+
+	topodatapb "vitess.io/vitess/go/vt/proto/topodata"
+	vschemapb "vitess.io/vitess/go/vt/proto/vschema"
+	"vitess.io/vitess/go/vt/srvtopo/srvtopotest"
+	"vitess.io/vitess/go/vt/topo"
+	"vitess.io/vitess/go/vt/topo/memorytopo"
+)
+
+var (
+	stockCell      = "some-cell"
+	stockCtx       = context.Background()
+	stockFilters   = []string{"bar", "baz"}
+	stockKeyspaces = map[string]*topodatapb.SrvKeyspace{
+		"foo": &topodatapb.SrvKeyspace{ShardingColumnName: "foo"},
+		"bar": &topodatapb.SrvKeyspace{ShardingColumnName: "bar"},
+		"baz": &topodatapb.SrvKeyspace{ShardingColumnName: "baz"},
+	}
+	stockVSchema = &vschemapb.SrvVSchema{
+		Keyspaces: map[string]*vschemapb.Keyspace{
+			"foo": &vschemapb.Keyspace{Sharded: true},
+			"bar": &vschemapb.Keyspace{Sharded: true},
+			"baz": &vschemapb.Keyspace{Sharded: false},
+		},
+	}
+)
+
+func newFiltering(filter []string) (*topo.Server, *srvtopotest.Server, Server) {
+	testServer := srvtopotest.New()
+
+	testServer.TopoServer = memorytopo.NewServer(stockCell)
+	testServer.SrvKeyspaceNames = []string{"foo", "bar", "baz"}
+	testServer.SrvKeyspace = &topodatapb.SrvKeyspace{ShardingColumnName: "test-column"}
+	testServer.WatchedSrvVSchema = stockVSchema
+
+	filtering, _ := NewKeyspaceFilteringServer(testServer, filter)
+	return testServer.TopoServer, testServer, filtering
+}
+
+func TestFilteringServerHandlesNilUnderlying(t *testing.T) {
+	got, gotErr := NewKeyspaceFilteringServer(nil, []string{})
+	if got != nil {
+		t.Errorf("got: %v, wanted: nil server", got)
+	}
+	if gotErr != ErrNilUnderlyingServer {
+		t.Errorf("Bad error returned: got %v wanted %v", gotErr, ErrNilUnderlyingServer)
+	}
+}
+
+func TestFilteringServerReturnsUnderlyingServer(t *testing.T) {
+	want, _, f := newFiltering(nil)
+	got := f.GetTopoServer()
+	if !reflect.DeepEqual(got, want) {
+		t.Errorf("Got incorrect topo.Server from FilteringServer: expected %p but got %p", want, got)
+	}
+}
+
+func doTestGetSrvKeyspaceNames(
+	t *testing.T,
+	f Server,
+	cell string,
+	want []string,
+	wantErr error,
+) {
+	got, gotErr := f.GetSrvKeyspaceNames(stockCtx, cell)
+
+	if !reflect.DeepEqual(got, want) {
+		t.Errorf("GetSrvKeyspaceNames failed: want %v, got %v", want, got)
+	}
+	if wantErr != gotErr {
+		t.Errorf("GetSrvKeyspaceNames returned incorrect error: want %v, got %v", wantErr, gotErr)
+	}
+}
+
+func TestFilteringServerGetSrvKeyspaceNamesFiltersKeyspaces(t *testing.T) {
+	_, _, f := newFiltering(stockFilters)
+	doTestGetSrvKeyspaceNames(t, f, stockCell, stockFilters, nil)
```

Comment:
This is slightly pedantic, but for completeness we might want to ensure that it returns an empty array (and not nil) if there is no intersection between the filter set and the keyspaces.